### PR TITLE
Add clustering algorithm

### PR DESCRIPTION
**Pull request by shakareem**
- Add clustering algorithm implemented by jetbrains-research.

- Add tests for it.

- Comply with the licence.

### CODE REVIEW

File: LICENSE.md [hunk 1]
```diff
@@ -19,3 +19,198 @@ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
...
+---
+
+All files in the package model.algorithms.clustering.implementation are provided under the following license:
+
+                                 Apache License
+                           Version 2.0, January 2004
+                        http://www.apache.org/licenses/
+
+TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTION
+
```
Comment (qrutyy) [inline LICENSE.md:32]: Do we really need to mention another license??????

**Review by qrutyy — CHANGES_REQUESTED** [review]

File: LICENSE.md (hunk 1 above)
Comment (shakareem) [reply LICENSE.md:32]: As i understood the rules, yes

Comment (qrutyy) [reply LICENSE.md:32]: Hm, ive never seen such "double-license". As I also understood, MIT is compatible with Apache 2.0, so mb it doesnt require doubling the license.
I've checked some links, if you have a proof, please send it. 
[using-code-under-apache-2-0-license-in-my-program-using-mit-license](https://softwareengineering.stackexchange.com/questions/297217/using-code-under-apache-2-0-license-in-my-program-using-mit-license)

**Review by qrutyy — APPROVED** [review]

**Review by spisladqo — APPROVED** [review]
I hope we will not get strike

**Pull request merged**